kolibri/plugins/learn/assets/src/modules/topicsTree/handlers.js

```diff
@@ -78,23 +79,25 @@ function _getChildren(id, topic, skip) {
   };
 }
 
-function _handleTopicRedirect(store, route, children, id, skipped) {
+function _handleTopicRedirect(next, route, children, id, skipped) {
   if (!children.some(c => !c.is_leaf) && route.name !== PageNames.TOPICS_TOPIC_SEARCH) {
     // if there are no children which are not leaf nodes (i.e. they have children themselves)
     // then redirect to search results
-    router.replace({
+    next({
       name: PageNames.TOPICS_TOPIC_SEARCH,
       params: { ...route.params, id },
       query: route.query,
+      replace: true,
     });
-    store.commit('SET_PAGE_NAME', PageNames.TOPICS_TOPIC_SEARCH);
+    return true;
   } else if (skipped) {
     // If we have skipped down the topic tree, replace to the new top level topic
-    router.replace({ name: route.name, params: { ...route.params, id }, query: route.query });
+    next({ name: route.name, params: { ...route.params, id }, query: route.query, replace: true });
+    return true;
   }
 }
 
-function _loadTopicsTopic(store, { route, baseurl } = {}) {
+function _loadTopicsTopic(next, { route, baseurl, shouldResolve } = {}) {
```

Comment:
This is not about this function specifically, but more that as I am reading through these handlers, I am realizing that despite reading through what I have so far, I do not have a good sense of when we are going to start needing to use this new `next` functionality. Will it be always, going forward? or is this a work around to deal with existing code, that we can approach differently going forward?

Reply:
One thing it allows is to properly redirect without risk of duplicate navigation errors or other errors from trying to navigate using the router while in a navigation guard.

I implemented it here because it was necessary to wait for the fetchChannel async call before we know if it's OK to complete the navigation or not.